crates/turbopack-ecmascript/src/analyzer/graph.rs

```diff
@@ -1410,6 +1410,37 @@ impl VisitAstPath for Analyzer<'_> {
         }
     }
 
+    fn visit_for_of_stmt<'ast: 'r, 'r>(
+        &mut self,
+        n: &'ast ForOfStmt,
+        ast_path: &mut swc_core::ecma::visit::AstNodePath<'r>,
+    ) {
+        let mut array = self.eval_context.eval(&n.right);
+        if let JsValue::Variable(id) = &array {
+            if let Some(value) = self.data.values.get(id) {
+                array = value.clone();
+            }
+        }
```

Comment:
You shouldn't read from `values` here. This should only happen during `linking`. Imagine this example:

``` js
let y;

x();

for(const z of y) {

}


function x() {
  y = [1,2,3];
}
```

In this case `values` is not set yet while visiting the for of loop.

---

Instead you need to keep the `Variable` and have some operation that converts an array to all iterated values of it: `JsValue::Iterated`.
Now you can define a built-in transform for `JsValue::Iterated(JsValue::Array(...))` to `JsValue::Alternatives`.